src/main/scala/org/apache/spark/sql/execution/datasources/greenplum/GreenplumUtils.scala

```diff
@@ -60,71 +61,174 @@ object GreenplumUtils extends Logging {
+      val partNum = df.rdd.getNumPartitions
+
+      df.foreachPartition { rows =>
+        copyPartition(rows, options, schema, tempTable, accumulator)
       }
-      (values.mkString(options.delimiter) + "\n").getBytes("UTF-8")
-    }
 
-    def convertValue(str: String): String = {
-      assert(options.delimiter.length == 1, "The delimiter should be a single character.")
-      val delimiter = options.delimiter.charAt(0)
-      str.flatMap {
-        case '\\' => "\\\\"
-        case '\n' => "\\n"
-        case '\r' => "\\r"
-        case `delimiter` => s"\\$delimiter"
-        case c => s"$c"
+      if (accumulator.value == partNum) {
+        val dropTbl = s"DROP TABLE IF EXISTS ${options.table}"
+        executeStatement(conn, dropTbl)
+
+        val renameTempTbl = s"ALTER TABLE $tempTable RENAME TO ${options.table}"
+        executeStatement(conn, renameTempTbl)
+      } else {
+        val dropTempTbl = s"DROP TABLE $tempTable"
+        executeStatement(conn, dropTempTbl)
+
+        throw new PartitionCopyFailureException(
+          s"""
+          | Job aborted for that there are some partitions failed to copy data to greenPlum:
+          | Total partitions is: ${partNum}, and successful partitions is: ${accumulator.value}.
+          | You can retry again.
+          """.stripMargin)
       }
+    } finally {
+      closeConnSilent(conn)
     }
+  }
 
+  /**
+   * https://www.postgresql.org/docs/9.2/sql-copy.html
+   *
+   * Copy data to greenplum in these cases, which need update origin gptable.
+   * 1. Overwrite an existed gptable, which is a CascadingTruncateTable.
+   * 2. Append data to a gptable.
+   *
+   * When transcationOn option is true, we will coalesce the dataFrame to one partition,
+   * and the copy operation for each partition is atomic.
+   *
+   * @param df the [[DataFrame]] will be copy to the Greenplum
+   * @param schema the table schema in Greemnplum
+   * @param options Options for the Greenplum data source
+   */
+  def nonTransactionalCopy(
+      df: DataFrame,
+      schema: StructType,
+      options: GreenplumOptions): Unit = {
     df.foreachPartition { rows =>
-      val conn = JdbcUtils.createConnectionFactory(options)()
-      val copyManager = new CopyManager(conn.asInstanceOf[BaseConnection])
+      copyPartition(rows, options, schema, options.table)
+    }
+  }
+
+  /**
+   * Copy a partition's data to a gptable.
+   *
+   * @param rows rows of a partition will be copy to the Greenplum
+   * @param options Options for the Greenplum data source
+   * @param schema the table schema in Greemnplum
+   * @param tableName the tableName, to which the data will be copy
+   * @param accumulator account for recording the successful partition num
+   */
+  def copyPartition(
+      rows: Iterator[Row],
+      options: GreenplumOptions,
+      schema: StructType,
+      tableName: String,
+      accumulator: LongAccumulator = null): Unit = {
```

Comment:
accumulator: LongAccumulator = None

then later we can call `accumulator. foreach(...)`

Reply:
done.